File: p2p_library/src/main/java/com/tambapps/p2p/fandem/util/IPUtils.java
package com.tambapps.p2p.fandem.util;

import java.io.IOException;
import java.net.InetAddress;
import java.net.NetworkInterface;
import java.net.ServerSocket;
import java.net.SocketException;
import java.util.Collections;
import java.util.List;

public class IPUtils {

    /**
     * from https://stackoverflow.com/questions/6064510/how-to-get-ip-address-of-the-device-from-code
     * @return return the ip address of the device
     */
    public static InetAddress getIPAddress() throws SocketException {
        List<NetworkInterface> interfaces = Collections.list(NetworkInterface.getNetworkInterfaces());
        for (NetworkInterface intf : interfaces) {
            List<InetAddress> addrs = Collections.list(intf.getInetAddresses());
            for (InetAddress addr : addrs) {
                if (!addr.isLoopbackAddress()) {
                    if (addr.getHostAddress().indexOf(':')<0) { //is ipv4
                        return addr;
                    }
                }
            }
        }
        return null;
    }

    /**
     * Get an available port
     * @param inetAddress the address of the host
     * @return an available port
     */
    public static int getAvalaiblePort(InetAddress inetAddress) {
        int port = 8081;
        while (port < 65536) {
            try (ServerSocket serverSocket = new ServerSocket(port,0, inetAddress)) {
                //tests that port available
            } catch (IOException e) {
                port++;
                continue;
            }
            return port;
        }
        throw new RuntimeException("No available port was found");
    }

}

File: p2p_library/src/main/kotlin/com/tambapps/p2p/fandem/util/IPUtils.kt
package com.tambapps.p2p.fandem.util

import com.tambapps.p2p.fandem.Peer
import com.tambapps.p2p.fandem.exception.NoPortAvailableException
import java.io.IOException
import java.net.InetAddress
import java.net.NetworkInterface
import java.net.ServerSocket
import java.util.Collections

object IPUtils {
  /**
   * from https://stackoverflow.com/questions/6064510/how-to-get-ip-address-of-the-device-from-code
   * @return return the ip address of the device
   */
  @JvmStatic
  @get:Throws(IOException::class)
  val ipAddress: InetAddress
    get() {
      val interfaces: List<NetworkInterface> = Collections.list(NetworkInterface.getNetworkInterfaces())
      for (intf in interfaces) {
        val addrs: List<InetAddress> = Collections.list(intf.inetAddresses)
        for (addr in addrs) {
          if (!addr.isLoopbackAddress) {
            if (addr.hostAddress.indexOf(':') < 0) { //is ipv4
              return addr
            }
          }
        }
      }
      throw IOException("Couldn't find IP")
    }

  @JvmStatic
  fun getIpAddressQuietly(): InetAddress? {
    return try {
      ipAddress
    } catch (e: IOException) {
      null
    }
  }

  /**
   * Get an available port
   * @param inetAddress the address of the host
   * @return an available port
   */
  @JvmStatic
  fun getAvailablePort(inetAddress: InetAddress?): Int {
    var port = Peer.DEFAULT_PORT
    while (port < Peer.DEFAULT_PORT + 16 * 16) {
      try {
        ServerSocket(port, 0, inetAddress).use { }
      } catch (e: IOException) {
        port++
        continue
      }
      return port
    }
    throw NoPortAvailableException("No available port was found")
  }

  /**
   * Returns a well formatted string of the given ip
   * @param inetAddress the address
   * @return a well formatted string of the given ip
   */
  @JvmStatic
  fun toString(inetAddress: InetAddress): String {
    return inetAddress.hostAddress.replace("/", "")
  }

  /**
   * Returns the hex string of the given ip
   * @param inetAddress the address
   * @return the hex string of the given ip
   */
  @JvmStatic
  fun toHexString(inetAddress: InetAddress): String {
    return toString(inetAddress).split(".").joinToString(prefix = "", postfix = "", separator = "") { s -> toHexString(s) }
  }

  @JvmStatic
  fun toHexString(s: String): String {
    return toHexString(s.toInt())
  }

  @JvmStatic
  fun toHexString(i: Int): String {
    val  n = i.toString(16)
    return (if (n.length == 1) "0$n" else n).toUpperCase()
  }
}